### PR TITLE
 Implement GenesisBlock() to Grab the Block from Memory DB

### PR DESCRIPTION
**Pull request by IvanTheGreatDev**
This addresses #349

### CODE REVIEW

File: beacon-chain/blockchain/core.go [hunk 1]
```diff
@@ -85,7 +85,7 @@ func NewBeaconChain(db ethdb.Database) (*BeaconChain, error) {
 
 // GenesisBlock returns the canonical, genesis block.
 func (b *BeaconChain) GenesisBlock() (*types.Block, error) {
-	return types.NewGenesisBlock()
+	return b.GetGenesisBlock()
```
[the remark this answers is not in the document]
Comment (IvanTheGreatDev) [reply beacon-chain/blockchain/core.go]: Good point, I'll change that

File: beacon-chain/blockchain/core.go [hunk 2]
```diff
@@ -140,6 +140,24 @@ func (b *BeaconChain) IsEpochTransition(slotNumber uint64) bool {
...
+	if err != nil {
+		return nil, err
+	}
+	if genesisExists {
+		bytes, _ := b.db.Get([]byte("genesis"))
+		block := &pb.BeaconBlockResponse{}
+		if err := proto.Unmarshal(bytes, block); err != nil {
+			return nil, err
+		}
+		// TODO: Add more default fields.
```
[the remark this answers is not in the document]
Comment (IvanTheGreatDev) [reply beacon-chain/blockchain/core.go]: 👍

File: beacon-chain/blockchain/core.go [hunk 3]
```diff
@@ -85,7 +85,19 @@ func NewBeaconChain(db ethdb.Database) (*BeaconChain, error) {
 
 // GenesisBlock returns the canonical, genesis block.
 func (b *BeaconChain) GenesisBlock() (*types.Block, error) {
-	return types.NewGenesisBlock()
+	genesisExists, err := b.db.Has([]byte("genesis"))
+	if err != nil {
+		return nil, err
+	}
+	if genesisExists {
+		bytes, _ := b.db.Get([]byte("genesis"))
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/core.go]: Please don’t dump this error. All errors must be handled.

**Comment by prestonvanloon** [comment]
@rawfalafel
What is the plan for the genesis block? 
How does the genesis block initially get onto disk?
Is there a separate issue tracking that or is it meant to be a requirement of #349?

File: beacon-chain/blockchain/core.go (hunk 3 above)
Comment (IvanTheGreatDev) [reply beacon-chain/blockchain/core.go]: Error handled 👍

File: beacon-chain/types/block.go [hunk 4]
```diff
@@ -31,18 +31,7 @@ func NewBlock(data *pb.BeaconBlockResponse) (*Block, error) {
 		return nil, errors.New("invalid block data, parent hash should be 32 bytes")
 	}
 
-	return &Block{data}, nil
-}
-
-// NewGenesisBlock returns the canonical, genesis block for the beacon chain protocol.
-func NewGenesisBlock() (*Block, error) {
```
Comment (nisdas) [inline beacon-chain/types/block.go]: why was this removed ?

File: beacon-chain/blockchain/core_test.go [hunk 5]
```diff
@@ -66,6 +67,35 @@ func TestNewBeaconChain(t *testing.T) {
...
+	if err := db.DB().Put([]byte("genesis"), bytes); err != nil {
+		t.Errorf("unable to save key value of genesis: %v", err)
+	}
+
+	genesisBlock, err := beaconChain.GenesisBlock()
+	if err != nil {
+		t.Errorf("unable to get key value of genesis: %v", err)
+	}
+
+	if time, _ := genesisBlock.Timestamp(); time.Second() != 40 {
```
Comment (nisdas) [inline beacon-chain/blockchain/core_test.go]: You are throwing away the error here , can you handle the error also

File: beacon-chain/types/block.go (hunk 4 above)
Comment (prestonvanloon) [reply beacon-chain/types/block.go]: Yeah, that is what I am wondering and prompted my question for @rawfalafel. How do we get the initial genesis before it's on disk?

Comment (rawfalafel) [reply beacon-chain/types/block.go]: One client in the network creates the genesis block and broadcasts it to everyone else. Every other client fetches and stores the block during initial sync. Since initial sync isn't done yet, this method should be called in `NewBeaconChain`, right next to active/crystallized state initialization.

File: beacon-chain/types/block.go [hunk 6]
```diff
@@ -31,18 +31,7 @@ func NewBlock(data *pb.BeaconBlockResponse) (*Block, error) {
...
-	genesisTime := time.Date(2018, time.July, 21, 12, 0, 0, 0, time.UTC)
```
Comment (rawfalafel) [inline beacon-chain/types/block.go]: Can you set this time to the current time?

Comment (IvanTheGreatDev) [reply beacon-chain/types/block.go]: Was about to say, it was weird to have it be a static date. Will do 👍

File: beacon-chain/blockchain/core.go [hunk 7]
```diff
@@ -49,11 +49,29 @@ func NewBeaconChain(db ethdb.Database) (*BeaconChain, error) {
 	if err != nil {
 		return nil, err
 	}
+	hasGenesis, err := db.Has([]byte("genesis"))
+	if err != nil {
+		return nil, err
+	}
+	if !hasGenesis {
+		genesisBlock, err := types.NewGenesisBlock()
```
Comment (terenc3t) [inline beacon-chain/blockchain/core.go]: Add a `log.Info` for something like `"No genesis block found on disk, initializing genesis block"`

**Comment by prestonvanloon** [comment]
So the tests are failing because it expects block slot 1,000,000 to be between genesis block time and `time.Now()`. Since this PR changes genesis block time to 1970, this condition is no longer met. 

This test would eventually start failing anyway as 1,000,000 slots have passed since the genesis block time. 

I recommend adding a `Clock` interface for testing, rather than relying on time.Now().
Inspired by this post on [StackOverflow](https://stackoverflow.com/questions/18970265/is-there-an-easy-way-to-stub-out-time-now-globally-in-golang-during-test).

1) Add a `Clock` interface that has a `Now()` method.
2) Provide a `RealClock` that implements the `Clock` interface by returning the result of `time.Now()` for the `Now()` method. Provide this clock as a package variable so it can be overridden in test. 
3) In the core_test.go, provide a test clock that you can use to manipulate the behavior of `Now()` to satisfy the scenario of the test. In this case, we expect the current time to be some time before the time that slot 1,000,000 starts.

**Comment by terenc3t** [comment]
@IvanTheGreatDev test is still failing

**Comment by prestonvanloon** [comment]
This looks good to me. Let's improve the coverage in another pull request.

**Comment by prestonvanloon** [comment]
@IvanTheGreatDev please follow up in another pull request to increase the coverage of this code.

**Comment by IvanTheGreatDev** [comment]
Will do! @prestonvanloon
